proposals/oci-support-updates.md

```diff
@@ -0,0 +1,95 @@
+
+
+## 2. Support for provenance files
+
+Helm currently has the ability to verify package signatures, assuming the presence of a file suffixed with `.prov` sitting next to a chart `.tgz` in a repository.
+
+Support for this method of signature validation should be carried over into OCI storage. As the format of the OCI manifest is custom to Helm, Helm can also choose to modify the resulting manifest on upload when a chart is being signed (e.g. `helm push --sign`).
+
+As far as the low-level details, the `.prov` file will simply be stored as a second layer on the manifest. If running `helm pull --verify oci://...`, the second layer (layer at index 1) will be assumed to be the provenance file. The first layer is always the chart `.tgz` itself.
+
+Here is an example of what a manifest might look like with a provenance file attached:
+
+```
+{
+  "schemaVersion": 2,
+  "config": {
+    "mediaType": "application/vnd.cncf.helm.config.v1+json",
+    "digest": "sha256:8ec7c0f2f6860037c19b54c3cfbab48d9b4b21b485a93d87b64690fdb68c2111",
+    "size": 117
+  },
+  "layers": [
+    {
+      "mediaType": "application/tar+gzip",
+      "digest": "sha256:1b251d38cfe948dfc0a5745b7af5ca574ecb61e52aed10b19039db39af6e1617",
+      "size": 2487
+    },
+    {
+      "mediaType": "application/pgp-signature",
+      "digest": "sha256:3e207b409db364b595ba862cdc12be96dcdad8e36c59a03b7b3b61c946a5741a",
+      "size": 643
+    }
+  ]
+}
+
+```
+
+If `application/pgp-signature` is not appropriate based the contents of a `.prov` file (e.g. it does not meet the IANA requirements), a custom media type will be used such as `application/vnd.cncf.helm.provenance.v1`.
+
+## 3. Chart versions == OCI reference tags
+
+To keep things simple, the version of a chart will be 1-to-1 with the tag used on registry references. Arbitrary tags will not be supported for initial release. As time goes on, support for arbitrary tags may be reintroduced.
+
+This also means that tags are no longer necessary to be provided on the command-line in the form `<ref>:<tag>`. Some users feel strongly to keep this to match the UX of Docker and other tools, so perhaps this feature is kept and treated the same as using the `--version` flag. Thoughts?
+
+The latest sematic version is able to be determined from a registry by listing all tags. To specify a specific tag/version (or if a registry does not support tag listing), users can provide `--version <version>`.
+
+## 4. Chart names == OCI reference basenames
+
+Again, to keep things simple, the basename (the last segment of the URL path) on a registry reference should be equivalent to the chart name. 
+
+For example, given a chart with the name `pepper` and the version `1.2.3`, users may run a command such as the following:
+
+```
+$ helm push pepper/ oci://r.myreg.io/mycharts
```

Comment:
Looking at https://github.com/helm/helm/issues/7032, I think this is incorrect or needs more care. This is doing two things: `helm package`, and `helm <upload>`. We probably should split those so `helm chart push` pushes a package you've already built, otherwise `helm push` needs to accept all the parameters `helm package` accepts.

For the on-liner use case we could perhaps add something like `helm package --push=<registry>` which is similar to how modern OCI Image builders work, e.g., [BuildKit](https://github.com/moby/buildkit/#imageregistry) (`--output type=image,name=docker.io/username/image,push=true`) and [Kaniko](https://github.com/GoogleContainerTools/kaniko/) (`--destination`). They don't also maintain a local image registry for execution, so their build commands are either filling a local cache, or actually end in an output to file or push to registry etc.

Reply:
On the flip side, `helm install ./` is a convenience method for `helm package` and `helm install`. We just don't provide additional feature flags to manipulate `helm package`. Those wishing to perform more advanced package methods (i.e. `--sign`) can use `helm package` instead.

Reply:
I realise this is slightly off-topic, but I'm unaware of a `helm push` issue for discussion. Would it be reasonable to (later) add support to `helm push` for non-OCI uploads, e.g. given a `http://` URL instead of an `oci://` URL, it trivially performs a HTTP PUT? The use-case here is for products like Artifactory which auto-generate the Helm Repository metadata when a chart is uploaded, to save carrying an extra tool around, i.e. `helm package && helm push` would be valid for both OCI Registries and ("smart") Chart Repositories with only a difference in the output URL.